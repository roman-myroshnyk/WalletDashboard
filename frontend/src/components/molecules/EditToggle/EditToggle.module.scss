.wrapper {
  display: flex;
  flex-direction: row-reverse;
  height: 100%;
}

.staleWrapper {
  position: absolute;

  display: flex;
  flex-direction: row-reverse;
  align-items: center;
  justify-content: space-between;
  width: 94px;
}

.editingWrapper {
  position: absolute;

  display: flex;
  flex-direction: row;
  gap: 12px;
  align-items: center;
  justify-content: space-between;
}
