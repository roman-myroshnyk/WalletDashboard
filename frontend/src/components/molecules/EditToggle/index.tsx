import { FC } from 'react';
import IconButton from '@/atoms/IconButton';
import CancelIcon from '@/icons/CancelIcon';
import ConfirmIcon from '@/icons/ConfirmIcon';
import EditIcon from '@/icons/EditIcon';
import styles from './EditToggle.module.scss';

export interface IProps {
  isEditing: boolean;
  disabled: boolean;
  onEdit: () => void;
  onCancel: () => void;
  onConfirm: () => void;
}
const EditToggle:FC<IProps> = ({
  isEditing,
  disabled,
  onEdit,
  onCancel,
  onConfirm,
}) => (
  <div className={styles.wrapper}>
    <div className={styles.staleWrapper}>
      <IconButton
        label="edit"
        isShown={!isEditing}
        disabled={disabled}
        onClick={onEdit}
      >
        <EditIcon
          label="edit"
          enterDelay={2}
          exitDelay={0}
        />
      </IconButton>
    </div>
    <div className={styles.editingWrapper}>
      <IconButton
        label="confirm"
        isShown={isEditing}
        disabled={disabled}
        onClick={onConfirm}
      >
        <ConfirmIcon
          label="confirm"
          enterDelay={2}
          exitDelay={0}
        />
      </IconButton>
      <IconButton
        label="cancel"
        isShown={isEditing}
        disabled={disabled}
        onClick={onCancel}
      >
        <CancelIcon
          label="cancel"
          enterDelay={1}
          exitDelay={1}
        />
      </IconButton>
    </div>
  </div>
);

export default EditToggle;
