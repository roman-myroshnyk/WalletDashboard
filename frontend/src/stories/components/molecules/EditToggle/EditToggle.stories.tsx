import { Meta, Story } from '@storybook/react/types-6-0';
import EditToggle, { IProps } from '@/molecules/EditToggle';

export default {
  title: 'molecules/EditToggle',
  component: EditToggle,
  argTypes: {
    isEditing: {
      controls: 'boolean',
    },
    disabled: {
      controls: 'boolean',
    },
    onEdit: {
      action: 'onEdit: edit button clicked',
    },
    onCancel: {
      action: 'onCancel: edit cancel button clicked',
    },
    onConfirm: {
      action: 'onConfirm: edit confirm button clicked',
    },
  },
} as Meta;

const Template:Story<IProps> = (args) => (
  <div
    style={{ width: '80px', height: '20px' }}
  >
    <EditToggle {...args} />
  </div>
);

export const Default = Template.bind({});

Default.args = {
  isEditing: false,
  disabled: false,
};
