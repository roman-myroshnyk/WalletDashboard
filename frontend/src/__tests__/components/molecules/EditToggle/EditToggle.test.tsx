import '.../../../mocks/framer-motion';
import { render, fireEvent } from '@testing-library/react';
import EditToggle from '@/molecules/EditToggle';

describe('molecules/EditToggle', () => {
  it('EditToggle should render stale state', () => {
    const handleEdit = jest.fn();
    const handleCancel = jest.fn();
    const handleConfirm = jest.fn();

    const { container } = render(
      <EditToggle
        isEditing={false}
        disabled={false}
        onEdit={handleEdit}
        onCancel={handleCancel}
        onConfirm={handleConfirm}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('EditToggle should render editing state', () => {
    const handleEdit = jest.fn();
    const handleCancel = jest.fn();
    const handleConfirm = jest.fn();

    const { container } = render(
      <EditToggle
        isEditing
        disabled={false}
        onEdit={handleEdit}
        onCancel={handleCancel}
        onConfirm={handleConfirm}
      />,
    );
    expect(container).toMatchSnapshot();
  });
  it('EditToggle all buttons must be disabled (editing state)', () => {
    const handleEdit = jest.fn();
    const handleCancel = jest.fn();
    const handleConfirm = jest.fn();

    const { getByTestId } = render(
      <EditToggle
        isEditing
        disabled
        onEdit={handleEdit}
        onCancel={handleCancel}
        onConfirm={handleConfirm}
      />,
    );
    const confirmButton = getByTestId('confirmButton') as HTMLButtonElement;
    const cancelButton = getByTestId('cancelButton') as HTMLButtonElement;
    expect(confirmButton.disabled).toBeTruthy();
    expect(cancelButton.disabled).toBeTruthy();
  });
  it('EditToggle all buttons must be disabled (stale state)', () => {
    const handleEdit = jest.fn();
    const handleCancel = jest.fn();
    const handleConfirm = jest.fn();

    const { getByTestId } = render(
      <EditToggle
        isEditing={false}
        disabled
        onEdit={handleEdit}
        onCancel={handleCancel}
        onConfirm={handleConfirm}
      />,
    );
    const editButton = getByTestId('editButton') as HTMLButtonElement;
    expect(editButton.disabled).toBeTruthy();
  });
  it('EditToggle onEdit should trigger (stale state)', () => {
    const handleEdit = jest.fn();
    const handleCancel = jest.fn();
    const handleConfirm = jest.fn();

    const { getByTestId } = render(
      <EditToggle
        isEditing={false}
        disabled={false}
        onEdit={handleEdit}
        onCancel={handleCancel}
        onConfirm={handleConfirm}
      />,
    );
    const editButton = getByTestId('editButton') as HTMLButtonElement;
    fireEvent.click(editButton);
    expect(handleEdit).toBeCalledTimes(1);
  });
  it('EditToggle onCancel should trigger (editing state)', () => {
    const handleEdit = jest.fn();
    const handleCancel = jest.fn();
    const handleConfirm = jest.fn();

    const { getByTestId } = render(
      <EditToggle
        isEditing
        disabled={false}
        onEdit={handleEdit}
        onCancel={handleCancel}
        onConfirm={handleConfirm}
      />,
    );
    const cancelButton = getByTestId('cancelButton') as HTMLButtonElement;
    fireEvent.click(cancelButton);
    expect(handleCancel).toBeCalledTimes(1);
  });
  it('EditToggle onConfirm should trigger (editing state)', () => {
    const handleEdit = jest.fn();
    const handleCancel = jest.fn();
    const handleConfirm = jest.fn();

    const { getByTestId } = render(
      <EditToggle
        isEditing
        disabled={false}
        onEdit={handleEdit}
        onCancel={handleCancel}
        onConfirm={handleConfirm}
      />,
    );
    const confirmButton = getByTestId('confirmButton') as HTMLButtonElement;
    fireEvent.click(confirmButton);
    expect(handleConfirm).toBeCalledTimes(1);
  });
});
