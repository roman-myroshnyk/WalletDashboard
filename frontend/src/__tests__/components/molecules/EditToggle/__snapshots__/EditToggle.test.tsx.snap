// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`molecules/EditToggle EditToggle should render editing state 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="staleWrapper"
    >
      <div />
    </div>
    <div
      class="editingWrapper"
    >
      <div>
        <button
          area-label="confirm"
          class="iconButton"
          data-testid="confirmButton"
          variants="[object Object]"
        >
          <svg
            animate="enter"
            aria-label="confirm"
            class="h-6 w-6"
            exit="exit"
            fill="none"
            height="24"
            initial="exit"
            stroke="currentColor"
            stroke-width="2"
            variants="[object Object]"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 13l4 4L19 7"
              stroke-linecap="round"
              stroke-linejoin="round"
              variants="[object Object]"
            />
          </svg>
        </button>
      </div>
      <div>
        <button
          area-label="cancel"
          class="iconButton"
          data-testid="cancelButton"
          variants="[object Object]"
        >
          <svg
            animate="enter"
            aria-label="cancel"
            class="h-6 w-6"
            data-testid="cancelIcon"
            exit="exit"
            fill="none"
            height="24"
            initial="exit"
            stroke="currentColor"
            stroke-width="2"
            variants="[object Object]"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18L18 6M6 12"
              stroke-linecap="round"
              stroke-linejoin="round"
              variants="[object Object]"
            />
            <path
              d="M6 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
              variants="[object Object]"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`molecules/EditToggle EditToggle should render stale state 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="staleWrapper"
    >
      <div>
        <button
          area-label="edit"
          class="iconButton"
          data-testid="editButton"
          variants="[object Object]"
        >
          <svg
            animate="enter"
            aria-label="edit"
            class="h-6 w-6"
            exit="exit"
            fill="none"
            height="24"
            initial="exit"
            stroke="currentColor"
            stroke-width="2"
            variants="[object Object]"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z"
              stroke-linecap="round"
              stroke-linejoin="round"
              variants="[object Object]"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="editingWrapper"
    >
      <div />
      <div />
    </div>
  </div>
</div>
`;
